src/backend/distributed/commands/table.c

```diff
@@ -593,59 +620,242 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
+{
+	Oid relationId = InvalidOid;
+	foreach_oid(relationId, relationIdList)
 	{
-		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
-											   alterTableStatement->missing_ok);
-		if (IsCitusTableType(rightRelationId, CITUS_LOCAL_TABLE))
+		if (OidIsValid(relationId) && !IsCitusTable(relationId))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ConvertPostgresLocalTablesToCitusLocalTables converts each postgres table
+ * involved in foreign keys to be defined by given alter table command and the
+ * other tables connected to them via a foreign key graph to citus local tables.
+ */
+static void
+ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	RangeVar *leftRelationRangeVar = alterTableStatement->relation;
+	List *relationRangeVarList = lappend(rightRelationRangeVarList, leftRelationRangeVar);
+
+	/*
+	 * To prevent deadlocks, sort the list before converting each postgres local
+	 * table to a citus local table.
+	 */
+	relationRangeVarList = SortList(relationRangeVarList, CompareRangeVarsByOid);
+
+	/*
+	 * Here we should operate on RangeVar objects since relations oid's would
+	 * change in below loop due to CreateCitusLocalTable.
+	 */
+	RangeVar *relationRangeVar;
+	foreach_ptr(relationRangeVar, relationRangeVarList)
+	{
+		List *commandList = alterTableStatement->cmds;
+		LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+		bool missingOk = alterTableStatement->missing_ok;
+		Oid relationId = RangeVarGetRelid(relationRangeVar, lockMode, missingOk);
+		if (!OidIsValid(relationId))
+		{
+			/*
+			 * As we are in preprocess, missingOk might be true and relation
+			 * might not exist.
+			 */
+			continue;
+		}
+		else if (IsCitusTable(relationId))
+		{
+			/*
+			 * relationRangeVarList has also reference and citus local tables
+			 * involved in this ADD FOREIGN KEY command. Moreover, even if
+			 * relationId was belonging to a postgres  local table initially,
+			 * we might had already converted it to a citus local table by cascading.
+			 */
+			continue;
+		}
+
+		/*
+		 * CopyErrorData() requires (CurrentMemoryContext != ErrorContext),
+		 * so we store CurrentMemoryContext here.
+		 */
+		MemoryContext savedMemoryContext = CurrentMemoryContext;
+		PG_TRY();
 		{
-			ErrorOutForFKeyBetweenPostgresAndCitusLocalTable(leftRelationId);
+			bool cascade = true;
+			CreateCitusLocalTable(relationId, cascade);
+		}
+		PG_CATCH();
+		{
+			MemoryContextSwitchTo(savedMemoryContext);
+
+			ErrorData *errorData = CopyErrorData();
+			FlushErrorState();
+
+			if (errorData->elevel != ERROR)
+			{
+				PG_RE_THROW();
+			}
+
+			/* extend error message with foreign key creation error */
+			StringInfo errorMessage = makeStringInfo();
+			appendStringInfo(errorMessage, "cannot create foreign key: %s",
+							 errorData->message);
```

Comment:
maybe get rid of as we are getting weird outputs like that:

>ERROR:  cannot create foreign key: cannot create foreign key constraint since foreign keys ....

Reply:
I think we could improve it as:
```
Creating foreign key failed to cascade because: %s
```